src/rdzstd.c

```diff
@@ -0,0 +1,197 @@
+                                   out_bufsize, ZSTD_getErrorName(ret));
+                        err = RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+                        goto done;
+                }
+        }
+
+done:
+
+        if (err) {
+                rd_free(decompressed);
+        }
+
+        return err;
+}
+
+rd_kafka_resp_err_t
+rd_kafka_zstd_compress (rd_kafka_broker_t *rkb, int comp_level,
+                       rd_slice_t *slice, void **outbuf, size_t *outlenp) {
+        ZSTD_CStream* cctx;
+        size_t r;
+        rd_kafka_resp_err_t err = RD_KAFKA_RESP_ERR_NO_ERROR;
+        size_t len = rd_slice_remains(slice);
+        ZSTD_outBuffer out;
+        ZSTD_inBuffer in;
+
+        *outbuf = NULL;
+        out.pos = 0;
+        out.size = ZSTD_compressBound(len);
+        out.dst = rd_malloc(out.size);
+        if (!out.dst) {
+                rd_rkb_dbg(rkb, MSG, "ZSTDCOMPR",
+                           "Unable to allocate output buffer "
+                           "(%"PRIusz" bytes): %s",
+                           out.size, rd_strerror(errno));
+                return RD_KAFKA_RESP_ERR__CRIT_SYS_RESOURCE;
+        }
+
+
+        cctx = ZSTD_createCStream();
+        if (!cctx) {
+                rd_rkb_dbg(rkb, MSG, "ZSTDCOMPR",
+                           "Unable to create ZSTD compression context");
+                err = RD_KAFKA_RESP_ERR__CRIT_SYS_RESOURCE;
+                goto done;
+        }
+
+        r = ZSTD_initCStream(cctx, comp_level);
+        if (ZSTD_isError(r)) {
+                rd_rkb_dbg(rkb, MSG, "ZSTDCOMPR",
+                           "Unable to begin ZSTD compression "
+                           "(out buffer is %"PRIusz" bytes): %s",
+                           out.size, ZSTD_getErrorName(r));
+                err = RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+                goto done;
+        }
+
+        while ((in.size = rd_slice_reader(slice, &in.src))) {
+                in.pos = 0;
+                r = ZSTD_compressStream(cctx, &out, &in);
+                if (unlikely(ZSTD_isError(r))) {
+                        rd_rkb_dbg(rkb, MSG, "ZSTDCOMPR",
+                                   "ZSTD compression failed "
+                                   "(at of %"PRIusz" bytes, with "
+                                   "%"PRIusz" bytes remaining in out buffer): "
+                                   "%s",
+                                   in.size, out.size - out.pos,
+                                   ZSTD_getErrorName(r));
+                        err = RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+                        goto done;
+                }
+
+                /* No space left in output buffer, but input isn't fully consumed */
+                if (in.pos < in.size) {
+                        err = RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+                        goto done;
+                }
+        }
+
+        rd_assert(rd_slice_remains(slice) == 0);
```

Comment:
Better to error out here instead of crashing since the input data is from an external source.